src/matcher/regex.js

```diff
@@ -14,7 +25,151 @@ import { noOpObj } from '@keg-hub/jsutils'
+  else if (end === '')
+    return `(${start}|${start}${middle})`
+  else
+    return `(${start}${end}|${start}${middle}${end})`
+}
+
+/**
+ * Gets the full text around an optional
+ * @param {Array<string>} match result of optional regex match
+ */
+const getFullOptionalText = (match) => {
+  const text = match.input
+  const precedingWord = getWordEndingAt(text, match.index)
+  return precedingWord + match[0]
+}
+
+/**
+ * Helper for `getParamRegex` to get the optional types regex
+ * @param {Array<string>} match result of optional regex match
+ * @return {string} the correct regex source text for an step definition
+ * optional part. This regex will be used for matching values in the
+ * feature step text.
+ */
+const getOptionalRegex = match => {
+  const optionalText = getFullOptionalText(match)
+  return toAlternateRegex(optionalText)
+}
+
+/**
+ * Returns regex for a given parameter type
+ * @param {string} type - cucumber-expression parameter type: float, int, word, or string
+ * @return {string} regex source text
+ */
+export const getParamRegex = type => {
+  switch(type) {
+    case 'float':
+      return RX_FLOAT.source
+    case 'int':
+      return RX_INT.source
+    case 'string':
+      return joinRegex(RX_DOUBLE_QUOTED, RX_SINGLE_QUOTED).source
+    case 'word':
+    default:
+      return RX_PARAM_REPLACE.source
+  }
+}
+
+/**
+ * Helper for `parseMatch` that gets the right regex
+ * for a step's dynamic part 
+ * @param {string} type - optional, alternate, or parameter
+ * @param {string} match = regex match results
+ */
+const getMatchRegex = (type, match) => {
+  const [ val, paramType ] = match
+
+  switch(type) {
+    case 'parameter':
+      return new RegExp(getParamRegex(paramType))
+    case 'optional':
+      return new RegExp(getOptionalRegex(match))
+    case 'alternate':
+      const pattern = `(${val.trim().replace(/\//g, '|')})`
+      return new RegExp(pattern)
+    default:
+      return null
+  }
+}
+
+/**
+ * Formats the regex match result into an object,
+ * with some computed values
+ * @param {Array} matchArr 
+ * @param {string} type 
+ */
+const parseMatch = (matchArr, type='other') => {
+  const val = matchArr[0]
+
+  return {
+    text: val.trim(),
+    index: matchArr.index,
+    input: matchArr.input,
+    regex: getMatchRegex(type, matchArr),
+    type,
+    ...(type === 'parameter' && { 
+      paramType: val.trim().replace(RX_MATCH_REPLACE, '')
+    })
+  }
+}
+
+/**
+ * Extracts all the dynamic parts to a step definition's match text
+ * @param {string} stepMatcher - step definition match text
+ * @return {Array<Object>} array of matches. See `parseMatch` for the structure.
+ * @example
+ * const parts = getRegexParts('I eat {int} apple(s)')
+ * result:
+ * [
+ *  { type: 'parameter', text: '{int}', regex: /-?[0-9]+/, paramType: 'int', ... } ,
+ *  { type: 'optional', text: 'apple(s)', regex: /(apple|apples)/, ... } ,
+ * ]
+ */
+export const getRegexParts = stepMatcher => {
```

Comment:
This is interesting. Not the way I would have done it, but also seems to work pretty well.
I'd be concerned about the sorting, but I have trust issues.

It seems you could have matching index's with the `optionals` and the `parameters`, but maybe I'm missing something?

Reply:
Hmm, they shouldn't? The optionals and parameters have different regex patterns that shouldn't overlap, but maybe I'm the one missing something.